extensions/amp-share-tracking/0.1/amp-share-tracking.js

```diff
@@ -15,43 +15,125 @@
  */
 
 import {isExperimentOn} from '../../../src/experiments';
-import {getService} from '../../../src/service';
+import {xhrFor} from '../../../src/xhr';
+import {viewerFor} from '../../../src/viewer';
 import {dev} from '../../../src/log';
 
 /** @private @const {string} */
 const TAG = 'amp-share-tracking';
 
-class AmpShareTracking extends AMP.BaseElement {
+export class AmpShareTracking extends AMP.BaseElement {
+  /**
+    * @return {boolean}
+    * @private
+    */
+  isExperimentOn_() {
+    return isExperimentOn(this.win, TAG);
+  }
+
+  /** @override */
+  isLayoutSupported(unusedLayout) {
+    return true;
+  }
 
   /** @override */
   buildCallback() {
-    this.isExperimentOn_ = isExperimentOn(this.win, TAG);
-    if (!this.isExperimentOn_) {
+    if (!this.isExperimentOn_()) {
       dev.warn(TAG, `TAG ${TAG} disabled`);
       return;
     }
+
+    /** @private @const {!Window} */
+    this.win_ = this.win;
+
+    this.vendorHref = this.element.getAttribute('data-href');
+    dev.fine(TAG, 'vendorHref: ', this.vendorHref);
+
+    this.shareTrackingFragments = Promise.all([
+      this.getIncomingFragment(),
+      this.getOutgoingFragment(this.vendorHref)]).then(results => {
+        const fragments = {};
+        fragments.incomingFragment = results[0];
+        fragments.outgoingFragment = results[1];
+        return fragments;
+      });
+
   }
-}
 
-/**
- * ShareTrackingService processes the incoming and outcoming url fragment
- */
-export class ShareTrackingService {
-  constructor(window) {
-    this.win = window;
+  /**
+   * Get the incoming share-tracking fragment from the viewer
+   * @return {!Promise<string>}
+   * @private
+   */
+  getIncomingFragment() {
+    return viewerFor(this.win_).getFragment().then(hash => {
+      if (!hash) {
+        return Promise.resolve();
+      }
+
+      if (hash.indexOf('.') == 0) {
+        const endIndex = hash.indexOf('&');
+        if (endIndex != -1) {
+          return Promise.resolve(hash.substr(1, (endIndex - 1)));
+        } else {
+          return Promise.resolve(hash.substr(1));
+        }
+      }
+      return Promise.resolve();
+    });
   }
-}
 
-/**
- * @param {!Window} window
- * @return {!ShareTrackingService}
- */
-export function installShareTrackingService(window) {
-  return getService(window, 'shareTrackingService', () => {
-    return new ShareTrackingService(window);
-  });
-}
+  /**
+   * Get an outgoing share-tracking fragment
+   * @param {string=} vendorUrl
+   * @return {!Promise<string>}
+   * @private
+   */
+  getOutgoingFragment(vendorUrl) {
+    if (vendorUrl) {
```

Comment:
I suggest we move the `vendorUrl` null check into `getOutgoingFragmentFromVendor`.
So here we can do:
```
return this.getOutgoingFragmentFromVendor(vendorUrl)
    .then(fragment => (fragment || this.getOutgoingFragmentRandomly());
```